packages/hdwallet-native/src/native.ts

```diff
@@ -7,10 +9,57 @@ import { MixinNativeBTCWallet, MixinNativeBTCWalletInfo } from "./bitcoin";
 import { MixinNativeETHWalletInfo, MixinNativeETHWallet } from "./ethereum";
 import { MixinNativeCosmosWalletInfo, MixinNativeCosmosWallet } from "./cosmos";
 import { MixinNativeBinanceWalletInfo, MixinNativeBinanceWallet } from "./binance";
+import type { NativeAdapterArgs } from "./adapter";
+
+export enum NativeEvents {
+  MNEMONIC_REQUIRED = "MNEMONIC_REQUIRED",
+  READY = "READY",
+}
+
+export class NativeHDWalletBase {
+  readonly #events: EventEmitter2;
+
+  constructor() {
+    this.#events = new EventEmitter2();
+  }
+
+  get events() {
+    return this.#events;
+  }
+
+  /**
+   * Wrap a function call that needs a mnemonic seed
+   * @param hasMnemonic - A
+   * @param callback
+   */
+  needsMnemonic<T>(hasMnemonic: boolean, callback: () => T): T {
+    if (hasMnemonic) {
+      return callback();
+    }
+
+    this.#events.emit(
+      NativeEvents.MNEMONIC_REQUIRED,
+      core.makeEvent({
+        message_type: NativeEvents.MNEMONIC_REQUIRED,
+        from_wallet: true,
+      })
+    );
+
+    throw new Error("Wallet is not initialized");
+  }
+
+  async needsMnemonicAsync<T>(hasMnemonic: boolean, callback: () => Promise<T>): Promise<T | null> {
+    try {
+      return await this.needsMnemonic(hasMnemonic, callback);
```

Comment:
And then here I feel like we should maybe not wrap in a try to allow the callback function to throw as it would be intended to instead of potentially eating the error.

Reply:
I'll take a look at this again. So far in my tests, any of these functions throwing an error was not being handled by axiom and was causing a crash, but I see your point about separating the error from `needsMnemonic` from the callback itself.

Reply:
Yeah, we should probably keep track of where axiom doesn't handle hdwallet errors correctly and actually address them